#3146466 - Add create remote entity feature.

shabananavas opened the pull request.

shabananavas committed:
Add create remote entity feature.

krystalcode commented on entity_sync.services.yml:
```diff
@@ -92,3 +92,10 @@ services:
       - { name: event_subscriber }
+
+  entity_sync.export.default_export_local_entity_terminate_subscriber:
+    class: Drupal\entity_sync\EventSubscriber\DefaultExportLocalEntityTerminate
```
**CHANGE REQUEST**

The ones that have to do with managing the imports/exports with the remote ID field should be called `Managed` instead of `Default`, like we did with `ManagedImportRemoteListFilters`. So let's call this `entity_sync.export.managed_local_entity_terminate_subscriber` (no need for `managed_**export**_local_entity` because of `entity_sync.**export**` prefix. And then `ManagedExportLocalEntityTerminate`.

krystalcode commented on src/EventSubscriber/DefaultExportLocalEntityTerminate.php:
```diff
@@ -0,0 +1,66 @@
+   * @param \Drupal\entity_sync\Event\TerminateOperationsEvent $event
+   *   The terminate operation event.
+   */
+  public function updateLocalEntity(TerminateOperationEvent $event) {
```
**CHANGE REQUEST**

Better name this `public function setRemoteId`.

shabananavas replied:
I've left it as updateLocalEntity now as we're also updating the changed field.

krystalcode commented on src/EventSubscriber/DefaultExportLocalEntityTerminate.php:
```diff
@@ -0,0 +1,66 @@
+      $context['local_entity'],
+      $event->getSync()
+    );
+  }
```
**CHANGE REQUEST**

Even though this is a response to an export operation, we are now **importing** the remote ID field into the local entity. We already have this method on the import field manager, let's make that method public and use it from there.

Also, call the `setRemoteChangedField` method as well so that we set the remote changed field as well. This will be useful to not import the entity when it shows up in the recently changed entities.

In our case with the Sync API where we don't have the `lastModified` field available in the response (yet - I've asked for it) we can just set the value of this property to the current request's time.

shabananavas committed:
Merge branch '3143534-export-local-entity-queue' into 3146466-export-local-entity-create

shabananavas committed:
Rename terminate subscriber.

shabananavas committed:
Merge branch '3143534-export-local-entity-queue' into 3146466-export-local-entity-create

shabananavas committed:
Take into account the remote entity changed timestamp for doing an update.

shabananavas committed:
Update function name.

shabananavas committed:
Merge branch '3143534-export-local-entity-queue' into 3146466-export-local-entity-create

shabananavas commented:
@krystalcode I believe I've addressed all the points we had talked about.

shabananavas committed:
Merge branch '3143534-export-local-entity-queue' into 3146466-export-local-entity-create

shabananavas committed:
:wq
Merge branch '3143534-export-local-entity-queue' into 3146466-export-local-entity-create

krystalcode commented:
Merged on Drupal.org without the code that prevents importing an export. A new PR should be created for that.
